querybook/server/logic/data_element.py

```diff
@@ -0,0 +1,156 @@
+from const.metastore import (
+    DataElementAssociationProperty,
+    DataElementAssociationTuple,
+    DataElementAssociationType,
+    DataElementTuple,
+)
+from lib.logger import get_logger
+from logic.user import get_user_by_name
+from models.data_element import DataElement, DataElementAssociation
+
+LOG = get_logger(__file__)
+
+
+@with_session
+def get_data_element_by_id(id: int, session=None):
+    return DataElement.get(id=id, session=session)
+
+
+@with_session
+def get_data_element_by_name(name: str, session=None):
+    return session.query(DataElement).filter(DataElement.name == name).first()
+
+
+@with_session
+def get_data_element_by_column_id(column_id: int, session=None):
+    results = (
+        session.query(DataElementAssociation)
+        .filter(DataElementAssociation.column_id == column_id)
+        .all()
+    )
+
+    if not results:
+        return None
+
+    # check if there are more than 1 association type
+    association_types = set([r.type for r in results])
+    if len(association_types) > 1:
+        LOG.error(
+            f"Column {column_id} has more than one data element associated with it"
+        )
+        return None
+
+    resp = {}
+    for row in results:
+        resp["type"] = row.type.value
+        resp[row.property_name.value] = (
+            row.data_element.to_dict() if row.data_element else row.primitive_type
+        )
+    return resp
+
+
+@with_session
+def create_or_update_data_element(
+    metastore_id: int, data_element_tuple: DataElementTuple, commit=True, session=None
+):
+    if data_element_tuple.created_by:
+        created_by = get_user_by_name(data_element_tuple.created_by, session=session)
+        if not created_by:
+            LOG.error(
+                f"Can't find created_by of data element {data_element_tuple.name}: {data_element_tuple.created_by}"
+            )
+
+        data_element_tuple.created_by = created_by or None
+
+    data_element = get_data_element_by_name(data_element_tuple.name, session=session)
+    fields = {**data_element_tuple._asdict(), "metastore_id": metastore_id}
+
+    if not data_element:
+        # create a new data element
+        data_element = DataElement.create(fields=fields, commit=commit, session=session)
+    else:
+        # update the data element
+        data_element = DataElement.update(
+            id=data_element.id, fields=fields, commit=commit, session=session
+        )
+
+    if commit:
+        session.commit()
+    else:
+        session.flush()
+
+
+@with_session
+def create_data_element_association_by_name(
+    data_element_name: str,
+    column_id: int,
+    association_type: DataElementAssociationType,
+    property_name: DataElementAssociationProperty,
+    primitive_type: str = None,
+    session=None,
+):
+    data_element = None
+    if data_element_name:
+        data_element = get_data_element_by_name(data_element_name, session=session)
+
+    if not data_element and not primitive_type:
+        LOG.error(
```

Comment:
just raise exception here and fail the whole flow with invalid values? Otherwise you might end up with partial data elements

Reply:
will check the return value of this function to determine if adding both key and value together, will not end up with partial data elements.

The reason of not raising exception to fail the whole flow is, adding data elements is just part of the whole table sync session, would like other parts to be finished even data elements can't be synced succesfully.